include/operations/Operation.hpp

```diff
@@ -96,6 +96,14 @@ namespace qc {
             return startQubit;
         }
 
+        [[nodiscard]] virtual std::vector<dd::Qubit> getUsedQubits() const {
+            auto usedQubits = targets;
+            for (auto control: controls) {
+                usedQubits.push_back(control.qubit);
+            }
+            return usedQubits;
+        }
```

Comment:
It is not sufficient to just define this in the base class without providing suitable overloads in the derived classes.
You can get around the overload for `ClassicControlledOperation` by using `getTargets()` and `getControls()` instead of `targets` and `controls` here.
For compound operations you have to loop through all operations and accumulate the respective qubits.
For non-unitary operations, it's just calling `getTargets()`.